{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: ipython in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (8.37.0)\n",
      "Requirement already satisfied: python-dotenv in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (1.1.1)\n",
      "Requirement already satisfied: requests in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (2.32.5)\n",
      "Requirement already satisfied: prompt_toolkit<3.1.0,>=3.0.41 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (3.0.51)\n",
      "Requirement already satisfied: stack_data in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (0.6.3)\n",
      "Requirement already satisfied: jedi>=0.16 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (0.19.2)\n",
      "Requirement already satisfied: matplotlib-inline in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (0.1.7)\n",
      "Requirement already satisfied: exceptiongroup in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (1.3.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (4.9.0)\n",
      "Requirement already satisfied: traitlets>=5.13.0 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (5.14.3)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (2.19.2)\n",
      "Requirement already satisfied: typing_extensions>=4.6 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (4.14.0)\n",
      "Requirement already satisfied: decorator in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from ipython) (5.2.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests) (2025.8.3)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests) (2.5.0)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests) (3.4.3)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests) (3.10)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from jedi>=0.16->ipython) (0.8.4)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from pexpect>4.3->ipython) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from prompt_toolkit<3.1.0,>=3.0.41->ipython) (0.2.13)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from stack_data->ipython) (3.0.0)\n",
      "Requirement already satisfied: pure-eval in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from stack_data->ipython) (0.2.3)\n",
      "Requirement already satisfied: executing>=1.2.0 in /Users/gerardnasseruncc/Library/Python/3.10/lib/python/site-packages (from stack_data->ipython) (2.2.0)\n",
      "\u001b[33mWARNING: You are using pip version 22.0.4; however, version 25.3 is available.\n",
      "You should consider upgrading via the '/Library/Frameworks/Python.framework/Versions/3.10/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\u001b[33m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install ipython python-dotenv requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "from dotenv import load_dotenv\n",
    "import urllib.parse\n",
    "import requests\n",
    "import json\n",
    "import csv\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## API Tutorial\n",
    "This document has two aims:\n",
    "\n",
    "- (1) To show you the process of registering and using an API application from Web of Science / Clarivate\n",
    "\n",
    "- (2) Utilize a script written by the author to efficiently retrieve results from multiple pages."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create an Account or Log In\n",
    "The first step is to register or log into https://developer.clarivate.com/ .\n",
    "\n",
    "![Step 1](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.40.48 PM.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Register the Application Step 1\n",
    "After you will register for an API application. Please click \"Register\".\n",
    "\n",
    "![Step 2](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.41.46 PM.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Register a New Application Step 2\n",
    "Then click on \"Register a new Application\" ...\n",
    "\n",
    "![Step 3](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.43.05 PM.png)\n",
    "\n",
    "... and fill out the form.\n",
    "\n",
    "![Step 4](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.43.45 PM.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Approval\n",
    "Once the registration is approved you can click on the Name link of your application as seen below.\n",
    "\n",
    "![Step 5](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.44.04 PM.png)\n",
    "\n",
    "Then click on the name if the API.\n",
    "\n",
    "![Step 6](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.44.37 PM.png)\n",
    "\n",
    "This will take you to a window where you can click the \"Try It\" button and you can then see the special UI for your registered application.\n",
    "\n",
    "![Step 7](./tutorial_screenshots/Screenshot%202025-09-03%20at%203.45.19 PM.png)\n",
    "\n",
    "The rest is play around and learn! This interface allows you to use the AP without code. It also shows you all teh advance search options and even allows you to perform a test search! This test search will tell you the curl command and show you the example output.\n",
    "\n",
    "However, it is not the most efficient way to use the API. Below is a notebook that will help you get more results faster. It is written in python and utilizes requests in place of curl. This helps keep it all in one language! Just remember to not exceed the limit of your subscription!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script Step 1\n",
    "This is where you will place your information such as API key, Database, and search phrase. All the abriviations are seen and explained in your individual APU UI. Please refer there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "find_dotenv() -> /Users/gerardnasseruncc/Desktop/All/Github/Project_Repositories/Prisma_Compliance_Resource/APIs/Web_of_Science_API_Resources_and_Tutorials/.env\n",
      "os.environ.get('page_max') = 10\n",
      "os.environ.get('PAGE_MAX') = None\n",
      "globals().get('page_max') = None\n",
      "(indoor spaces OR indoor environment OR indoor microbiome OR built environment OR office OR home OR apartment OR classroom OR building) AND (plant wall OR green wall OR phytoremediation system OR vertical planting OR indoor plant OR potted plant OR house plant OR ornamental plants OR plant intervention OR soil intervention OR hydroponic) AND (skin microbiome OR skin samples OR soil microbiome OR surface microbiome OR surface samples OR dust OR dust samples OR bioaerosols OR indoor air OR rhizosphere microbiome OR biofilter) AND (16s gene OR 16S rrna OR 16s sequencing OR 16s amplification OR 16s data OR 16s analysis OR 16s rna OR shotgun sequencing OR bacterial characterization OR bacterial composition OR bacterial analysis OR microbiome analysis OR microbial diversity OR microbiota OR metagenomics)\n",
      "10\n",
      "8646414b229139f84e5861bd43bd8b8fa6084fdb\n"
     ]
    }
   ],
   "source": [
    "# Load environment variables from .env file\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "import os\n",
    "print(\"find_dotenv() ->\", find_dotenv())   # path dotenv will use when load_dotenv() is called without args\n",
    "load_dotenv()  \n",
    "# Show both casings and raw os.environ entry\n",
    "print(\"os.environ.get('page_max') =\", os.environ.get('page_max'))\n",
    "print(\"os.environ.get('PAGE_MAX') =\", os.environ.get('PAGE_MAX'))\n",
    "# Also show if a python variable by that name exists in the kernel\n",
    "print(\"globals().get('page_max') =\", globals().get('page_max'))\n",
    "# Use the environment variable, but fall back to a placeholder if it's not set.\n",
    "# Replace 'YOUR_API_KEY' with your actual key if you are not using a .env file.\n",
    "API_KEY = os.environ.get('API_KEY', 'YOUR_API_KEY')\n",
    "database_code = os.environ.get('database_code')\n",
    "page_limit = int(os.environ.get('page_limit', 50))\n",
    "page_max = int(os.environ.get('page_max', 10))\n",
    "sort_field = os.environ.get('sort_field')\n",
    "ascending = os.environ.get('ascending')\n",
    "base_url = os.environ.get('base_url')\n",
    "search_phrase = os.environ.get('search_phrase')\n",
    "\n",
    "print(search_phrase)\n",
    "print(page_max)\n",
    "print(API_KEY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is a definition used to outline a repeatable task.\n",
    "# This function creates an encoded query string for a search phrase.\n",
    "def create_encoded_query(phrase):\n",
    "    \"\"\"Encodes a string for use in a URL.\"\"\"\n",
    "    return urllib.parse.quote(phrase)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script Step 2\n",
    "This step will build your url phrase that will be added to the base url. This creates the full request that will then be sent to the API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TS=((indoor spaces OR indoor environment OR indoor microbiome OR built environment OR office OR home OR apartment OR classroom OR building) AND (plant wall OR green wall OR phytoremediation system OR vertical planting OR indoor plant OR potted plant OR house plant OR ornamental plants OR plant intervention OR soil intervention OR hydroponic) AND (skin microbiome OR skin samples OR soil microbiome OR surface microbiome OR surface samples OR dust OR dust samples OR bioaerosols OR indoor air OR rhizosphere microbiome OR biofilter) AND (16s gene OR 16S rrna OR 16s sequencing OR 16s amplification OR 16s data OR 16s analysis OR 16s rna OR shotgun sequencing OR bacterial characterization OR bacterial composition OR bacterial analysis OR microbiome analysis OR microbial diversity OR microbiota OR metagenomics))\n"
     ]
    }
   ],
   "source": [
    "# Adjust this line to adjust your search fields.\n",
    "search_phrase_with_tag = f'TS=({search_phrase})'\n",
    "print(search_phrase_with_tag)\n",
    "\n",
    "params = {\n",
    "    'db': database_code,\n",
    "    'q': search_phrase_with_tag,\n",
    "    'limit': page_limit\n",
    "}\n",
    "\n",
    "if ascending == 'true':\n",
    "    params['sortField'] = f'{sort_field} A' # Note: API docs use a space not a + like in the UI.\n",
    "elif ascending == 'false':\n",
    "    params['sortField'] = f'{sort_field} D'\n",
    "else:\n",
    "    params['sortField'] = sort_field\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "headers = {\n",
    "    'accept': 'application/json',\n",
    "    'X-ApiKey': API_KEY\n",
    "}\n",
    "\n",
    "all_results = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script Step 3\n",
    "This section does the actual search, retrieving results from pages 1-X, where X is the page limit you defined earlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting page 1/10...\n",
      "  ... Request error on page 1: 400 Client Error: Bad Request for url: https://api.clarivate.com/apis/wos-starter/v1/documents?db=WOk&q=TS%3D%28%28indoor+spaces+OR+indoor+environment+OR+indoor+microbiome+OR+built+environment+OR+office+OR+home+OR+apartment+OR+classroom+OR+building%29+AND+%28plant+wall+OR+green+wall+OR+phytoremediation+system+OR+vertical+planting+OR+indoor+plant+OR+potted+plant+OR+house+plant+OR+ornamental+plants+OR+plant+intervention+OR+soil+intervention+OR+hydroponic%29+AND+%28skin+microbiome+OR+skin+samples+OR+soil+microbiome+OR+surface+microbiome+OR+surface+samples+OR+dust+OR+dust+samples+OR+bioaerosols+OR+indoor+air+OR+rhizosphere+microbiome+OR+biofilter%29+AND+%2816s+gene+OR+16S+rrna+OR+16s+sequencing+OR+16s+amplification+OR+16s+data+OR+16s+analysis+OR+16s+rna+OR+shotgun+sequencing+OR+bacterial+characterization+OR+bacterial+composition+OR+bacterial+analysis+OR+microbiome+analysis+OR+microbial+diversity+OR+microbiota+OR+metagenomics%29%29&limit=50&sortField=LD+A&page=1\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "if not base_url:\n",
    "    print(\"Error: 'base_url' is not set. Please check your .env file or environment variables.\")\n",
    "else:\n",
    "    for i in range(1, page_max + 1):\n",
    "        print(f\"Requesting page {i}/{page_max}...\")\n",
    "        # Set the page for the current iteration\n",
    "        params['page'] = i\n",
    "\n",
    "        try:\n",
    "            # Make the API call with a timeout and raise for status to catch HTTP errors\n",
    "            response = requests.get(base_url, headers=headers, params=params, timeout=30)\n",
    "            response.raise_for_status()\n",
    "        except requests.exceptions.RequestException as e:\n",
    "            print(f\"  ... Request error on page {i}: {e}\")\n",
    "            break\n",
    "\n",
    "        print(response.url)  # Print the full URL for debugging\n",
    "\n",
    "        try:\n",
    "            data = response.json()\n",
    "        except ValueError:\n",
    "            print(f\"  ... Failed to decode JSON on page {i}. Response text:\\n{response.text}\")\n",
    "            break\n",
    "\n",
    "        # The list of documents is in the 'hits' key of the JSON response\n",
    "        page_hits = data.get('hits', [])\n",
    "        if page_hits:\n",
    "            # Add the results from this page to our master list\n",
    "            all_results.extend(page_hits)\n",
    "            print(f\"  ... Success! Added {len(page_hits)} results.\")\n",
    "        else:\n",
    "            # Stop if a page returns no results\n",
    "            print(\"  ... No more results found. Stopping.\")\n",
    "            break\n",
    "\n",
    "        # small delay to avoid hitting rate limits\n",
    "        time.sleep(0.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script Step 4\n",
    "Finally, the list of JSONs is exported into a more human readable format, a csv. For down stream use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "if all_results:\n",
    "    publication_list = []\n",
    "\n",
    "    # Begin Loop\n",
    "    for record in all_results:\n",
    "        # Create Empty Dict\n",
    "        processed_record = {}\n",
    "\n",
    "        # Get UID\n",
    "        processed_record['UID'] = record.get('uid', 'N/A')\n",
    "\n",
    "        # Get Nested Identifier Information\n",
    "        identifiers_info = record.get('identifiers', {})\n",
    "        processed_record['DOI'] = identifiers_info.get('doi', 'N/A')\n",
    "\n",
    "        # Get Title\n",
    "        processed_record['Title'] = record.get('title', 'N/A').lower()\n",
    "\n",
    "        # Get the Data That is Nested\n",
    "        source_info = record.get('source', {}) # Get source, or an empty dict if it's missing\n",
    "        processed_record['Year'] = source_info.get('publishYear', 'N/A')\n",
    "        processed_record['Journal'] = source_info.get('sourceTitle', 'N/A')\n",
    "\n",
    "        # Get Processed Author Data\n",
    "        names_info = record.get('names', {})\n",
    "        authors_list = names_info.get('authors', []) # Get authors list, or an empty list\n",
    "        if authors_list:\n",
    "            # Create a list of just the author names\n",
    "            author_names = [author.get('displayName') for author in authors_list]\n",
    "            # Join them into a single string separated by a semicolon\n",
    "            processed_record['Authors'] = '; '.join(author_names)\n",
    "        else:\n",
    "            processed_record['Authors'] = 'N/A'\n",
    "\n",
    "        publication_list.append(processed_record)\n",
    "    publication_dataframe = pd.DataFrame(publication_list)\n",
    "    publication_dataframe.to_csv('wos_processed_results.csv', index=False)\n",
    "    print(f\"Data written to wos_processed_results.csv\")\n",
    "    print(pd.Timestamp.now())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
